cmd/dockerd/trap/trap.go

```diff
@@ -4,54 +4,38 @@ import (
 	"fmt"
 	"os"
 	"os/signal"
-	"sync/atomic"
 	"syscall"
 )
 
 // Trap sets up a simplified signal "trap", appropriate for common
 // behavior expected from a vanilla unix command-line tool in general
 // (and the Docker engine in particular).
 //
-//   - If SIGINT or SIGTERM are received, `cleanup` is called, then the process is terminated.
-//   - If SIGINT or SIGTERM are received 3 times before cleanup is complete, then cleanup is
-//     skipped and the process is terminated immediately (allows force quit of stuck daemon)
-//   - Ignore SIGPIPE events. These are generated by systemd when journald is restarted while
-//     the docker daemon is not restarted and also running under systemd.
-//     Fixes https://github.com/docker/docker/issues/19728
+// The first time a SIGINT or SIGTERM signal is received, `cleanup` is called in
+// a new goroutine.
+//
+// If SIGINT or SIGTERM are received 3 times, the process is terminated
+// immediately (allows force quit of stuck daemon).
 func Trap(cleanup func(), logger interface {
 	Info(args ...interface{})
 }) {
-	c := make(chan os.Signal, 1)
-	// we will handle INT, TERM, SIGPIPE here
-	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGPIPE)
+	c := make(chan os.Signal, 10)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```

Comment:
Maybe `signal.NotifyContext` and we can pass around the context?

Reply:
`signal.Notify` would still be needed to implement the "force-quit on three ^C" behaviour.